### ISSUE TITLE
🐞 <title>Fetching remote repository with SSH non-default port

### ISSUE THREAD

**Issue by pizza-prosciutto** [dagger/dagger#8638]
### What is the issue?

Fetching modules from a private repository using SSH fails when the repository uses a port other than 22.
Our privately hosted Bitbucket exposes port 7999 instead of the default 22.
Seems like the port in the URL is not taken into account.

### Dagger version

dagger v0.13.1 (registry.dagger.io/engine:v0.13.1) linux/amd64

### Steps to reproduce

Create a new module with `dagger init --sdk=go modules/hello`
Save it in a private repo which is accessible via SSH on a port other 7999
Try executing `dagger -m ssh://git@<private-bitbucket-server>:7999/<project>/dagger-demo.git/modules/hello functions`

### Log output

➜  /tmp dagger -m ssh://git@<private-bitbucket-server>:7999/<project>/dagger-demo.git/modules/hello functions
✔ connect 0.2s
✘ initialize 1m24.4s
! failed to get configured module: failed to get module ref kind: input: moduleSource resolve: failed to resolve git src to commit: failed to load cache key: error fetching default branch for repository ssh://xxxxx@<private-bitbucket-server>/<project>/dagger-demo.git: git error: exit status 128
  ✘ resolving module ref 1m24.4s
  ! failed to get configured module: failed to get module ref kind: input: moduleSource resolve: failed to resolve git src to commit: failed to load cache key: error fetching default branch for repository ssh://xxxxx@<private-bitbucket-server>/<project>/dagger-demo.git: git error: exit status 128
    ✘ moduleSource(refString: "ssh://git@<private-bitbucket-server>:7999/<project>/dagger-demo.git/modules/hello"): ModuleSource! 1m24.3s
    ! failed to resolve git src to commit: failed to load cache key: error fetching default branch for repository ssh://xxxxx@<private-bitbucket-server>/<project>/dagger-demo.git: git error: exit status 128
      ✘ cache request: git://<private-bitbucket-server>/<project>/dagger-demo.git 42.1s
      ! error fetching default branch for repository ssh://xxxxx@<private-bitbucket-server>/<project>/dagger-demo.git: git error: exit status 128

A new release of dagger is available: v0.13.1 → v0.13.3
To upgrade, see https://docs.dagger.io/install
https://github.com/dagger/dagger/releases/tag/v0.13.3
Error: failed to get configured module: failed to get module ref kind: input: moduleSource resolve: failed to resolve git src to commit: failed to load cache key: error fetching default branch for repository ssh://xxxxx@<private-bitbucket-server>/<project>/dagger-demo.git: git error: exit status 128
stderr:
ssh: connect to host <private-bitbucket-server> port 22: Connection refused
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

**Comment by jedevc** [comment]
cc @grouville

**Comment by grouville** [comment]
Hey @pizza-prosciutto,

Did you manage to get a workaround ? I don't think I'll have the bandwidth this week to fix it 👼 Keep me in touch 🙏

**Comment by pizza-prosciutto** [comment]
Hi @grouville 
No issues, thank you for taking care of it. We have moved the repository to GitHub as a workaround temporarily and it works nicely.

**Comment by fcs-vtl** [comment]
Hi there! We are facing the same issue using on on-premise private git repo listening on port 7999 when working with modules. We have had to put everything in a single module for the time being.

**Comment by Drilmo** [comment]
Hi, same issue on module dependencies.. 
https://discord.com/channels/707636530424053791/708371226174685314/1408525333162098808

Is a fix planned?
Dagger is a solution my client is considering, but this behavior is currently hindering adoption, given the complexity of the CI in place. If the modules cannot be in different repositories, it will not be possible (with a DevOps team of 50 people spread across several technical stacks, it would be complicated for maintenance if we had to use a single repository).

**Comment by jedevc** [comment]
Yeah, we need to update the logic here: https://github.com/dagger/dagger/blob/e035126279e30913831033156f0b9d544873f896/core/modulerefs.go#L137-L137

Essentially we should be using `sshutil` for parsing SCP strings, and not confuse them with `ssh://` URLs - I suspect this is where the issue is coming from (git URLs are weird).

I'll take a look.

**Comment by Drilmo** [comment]
> Yeah, we need to update the logic here:
> 
> [dagger/core/modulerefs.go](https://github.com/dagger/dagger/blob/e035126279e30913831033156f0b9d544873f896/core/modulerefs.go#L137-L137)
> 
> Line 137 in [e035126](/dagger/dagger/commit/e035126279e30913831033156f0b9d544873f896)
> 
>  func ParseGitRefString(ctx context.Context, refString string) (_ ParsedGitRefString, rerr error) { 
> Essentially we should be using `sshutil` for parsing SCP strings, and not confuse them with `ssh://` URLs - I suspect this is where the issue is coming from (git URLs are weird).
> 
> I'll take a look.

Thanks. Agreed on parsing only SCP-like refs with sshutil to avoid confusing them with ssh:// URLs, as you noted in ParseGitRefString() -> core/modulerefs.go:137.

From my side, the failing case is an explicit ssh://host:port URL: the “:443” is preserved by the endpoint but gets dropped when reconstructing cloneRef from RepoRoot.Root, so git falls back to port 22:
- endpoint with host:port is created, then cloneRef is rebuilt from RepoRoot.Root, losing the port ParseGitRefString() -> core/modulerefs.go:217
- GIT_SSH_COMMAND is forced and ignores ~/.ssh/config, so without the port in the URL ssh defaults to 22 getGitSSHCommand() ->util/gitutil/cli.go:308

Two complementary fixes:
- Keep the explicit port when building cloneRef for ssh://, e.g. replace the host part of RepoRoot.Root with the original host:port from endpoint.
- Optionally, propagate “-p &lt;port&gt;” into GIT_SSH_COMMAND when the URL includes a port.

For SCP-like parsing, there’s prior art via sshutil already used here util.gitutil.url -> util/gitutil/url.go:128.
